### PR TITLE
feature: add db-state dump command to `optics-cli`

### LINKED ISSUE #822: DB: have prover sync store information about update contents
Currently prover sync traverses from update to update, ingesting leaves

-  prover_sync should store metadata about tree structure, 
    - which leaves an update contains (from_index, to_index)
- homeindexer should store metadata about update events
    - block height, timestamp

any other good things to store @yourbuddyconner ?

### PR DESCRIPTION
**Pull request by ltchang2019**
- refactors `prove-cli` into `optics-cli` and adds `db-state` as subcommand (dumps state of db left by home indexer)
  - now uses `StructOpt` instead of `Clap`
- `db-state` command prints all `update.new_root`, the `block_number` of the update, and all contained `leaves`
- in `/rust/tools/optics-cli`, run `cargo run db-state --db-path <db_path> --home-name <home_name>`

Closes #822

Screen Shot 2021-09-30 at 10 04 59 PM
Screen Shot 2021-10-01 at 9 34 58 AM

### CODE REVIEW

**Review by prestwich — CHANGES_REQUESTED** [review]
separate out the core features from the cli feature

File: rust/tools/optics-cli/src/subcommands/prove.rs [hunk 1]
```diff
@@ -21,53 +23,68 @@ use once_cell::sync::OnceCell;
...
+#[derive(StructOpt, Debug)]
+pub struct ProveCommand {
+    /// Leaf to prove
+    #[structopt(long, required_unless = "leaf_index")]
+    leaf: Option<H256>,
 
     /// Leaf index to prove
-    #[clap(long)]
-    leaf_hash: Option<H256>,
+    #[structopt(long, required_unless = "leaf")]
```
Comment (prestwich) [inline rust/tools/optics-cli/src/subcommands/prove.rs:37]: oh that's nice

File: rust/tools/optics-cli/src/subcommands/db_state.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+                    output_map.insert((new_root, update_metadata.block_number), bucket);
+                }
+                // No more updates left
+                None => break,
+            }
+        }
+
+        let mut output_vec: Vec<_> = output_map.into_iter().collect();
+        output_vec.sort_by(|x, y| x.0 .1.cmp(&y.0 .1));
```
Comment (prestwich) [inline rust/tools/optics-cli/src/subcommands/db_state.rs:96]: add an explanation of this line

File: rust/tools/optics-cli/src/subcommands/db_state.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+        Ok(output_vec)
+    }
+
+    fn print_output(output_vec: Vec<((H256, u64), Vec<CommittedMessage>)>) {
```
Comment (prestwich) [inline rust/tools/optics-cli/src/subcommands/db_state.rs:101]: add an option to output to a json file

**Comment by ltchang2019** [comment]
Separated PR into 2 PRs
Closing in favor of #863 and #864

**Pull request closed without merging**